src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs

```diff
@@ -24,27 +24,68 @@
 // OTHER DEALINGS IN THE SOFTWARE.
 
 using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace Autofac.Extensions.DependencyInjection
 {
     internal class AutofacServiceProvider : IServiceProvider, ISupportRequiredService
     {
         private readonly IComponentContext _componentContext;
+        private List<OrderedServiceDescriptor> _originalDescriptors;
+        private HashSet<Type> _serviceTypes;
+        private MethodInfo _castMethodInfo;
 
-        public AutofacServiceProvider(IComponentContext componentContext)
+        public AutofacServiceProvider(IComponentContext componentContext, List<ServiceDescriptor> originalDescriptors)
         {
             _componentContext = componentContext;
+            _originalDescriptors = originalDescriptors.Select((d, i) => new OrderedServiceDescriptor
+            {
+                Index = i,
+                ImplementationInstance = d.ImplementationInstance,
+                ImplementationType = d.ImplementationType
+            }).ToList();
+            _serviceTypes = new HashSet<Type>(originalDescriptors.Select(d => d.ServiceType));
+            _castMethodInfo = typeof(Enumerable).GetTypeInfo().GetDeclaredMethod("Cast");
         }
 
         public object GetService(Type serviceType)
         {
-            return _componentContext.ResolveOptional(serviceType);
+            return GetOrdered(serviceType, _componentContext.ResolveOptional(serviceType));
```

Comment:
It's not sufficient to just reorder the return value of `IServiceProvider.GetService`. Constructor injection has to work the same way.

In the past, when we required support for a fallback `IServiceProvider`, we implemented a [simple IRegistrationSource](https://github.com/aspnet/DependencyInjection/blob/1.0.0-beta1/src/Microsoft.Framework.DependencyInjection.Autofac/ChainedRegistrationSource.cs) so services from the fallback provider could be resolved both via `GetService` and constructor injection. (The irony is not lost on me btw...)

I was initially hoping that the fix might be as simple as iterating of the descriptors in reverse in `AutofacRegistration.Register`, but seeing how the service descriptors rely on multiple registration sources themselves (such as for open generics), I now understand that it's not.